File: leetcode/43-multiply-strings/main.go
package main

import (
	"fmt"
	"strconv"
)

/*
	string conversion is not allowed
*/
func multiply0(num1 string, num2 string) string {
	a, _ := strconv.Atoi(num1)
	b, _ := strconv.Atoi(num2)
	c := a * b
	return strconv.Itoa(c)
}

/*
	ref: https://leetcode.com/problems/multiply-strings/discuss/17605/Easiest-JAVA-Solution-with-Graph-Explanation

	Time	O(M*N)
	Space	O(M+N)
	0 ms, faster than 100.00%
*/
func multiply(num1 string, num2 string) string {
	// init 0000000...the number of digits of any 2 numbers mulitplication must <= N+M digits
	res := []int{}
	for i := 0; i < len(num1)+len(num2); i++ {
		res = append(res, 0)
	}
	// calculate
	for i := len(num1) - 1; i >= 0; i-- {
		for j := len(num2) - 1; j >= 0; j-- {
			// position1: leading digit e.g. 91, 9 is at position1
			// position2: ending digit e.g. 91, 1 is at position1
			position1 := i + j
			position2 := i + j + 1
			// multiply the digits, it must gets 2 digits result
			x := int(num1[i] - '0')
			y := int(num2[j] - '0')
			// add the carray, which was stored in the last computation
			z := x*y + res[position2]
			// put them into the result array
			res[position1] += z / 10
			res[position2] = z % 10
		}
	}
	result := ""
	started := false
	for _, c := range res {
		if !started && c == 0 {
			continue
		}
		started = true
		result += strconv.Itoa(c)
	}
	if len(result) == 0 {
		return "0"
	}
	return result
}

func main() {
	/*
	   "498828660196"
	   "840477629533"
	*/

	fmt.Println(multiply("12", "345"))
	fmt.Println(multiply("99", "99"))
	fmt.Println(multiply("1", "12"))
	fmt.Println(multiply("1", "0"))
}

File: leetcode/43-multiply-strings/main.py
"""
    see ./idea.jpg

    Time	O(MN)
	Space	O(M+N)
    188 ms, faster than 38.41%
"""


class Solution(object):
    def multiply(self, num1, num2):
        """
        :type num1: str
        :type num2: str
        :rtype: str
        """
        if num1 == '0' or num2 == '0':
            return '0'
        M, N = len(num1), len(num2)
        digits = (M + N) * [0]
        for i in range(M):
            for j in range(N):
                _i = M - i - 1
                _j = N - j - 1
                temp = int(num1[_i]) * int(num2[_j])
                digits[i+j] += temp % 10
                digits[i+j+1] += temp//10
        for i in range(len(digits)-1):
            carry = digits[i]//10
            digits[i] = digits[i] % 10
            digits[i+1] += carry
        digits.reverse()
        res = ''
        for i in range(len(digits)):
            if res == '' and digits[i] == 0:
                continue
            res += str(digits[i])
        return res


s = Solution()

a = '123'
b = '456'
print(s.multiply(a, b))

a = '123'
b = '56'
print(s.multiply(a, b))

a = "498828660196"
b = "840477629533"
print(s.multiply(a, b))

a = '0'
b = '456'
print(s.multiply(a, b))
